compiler/rustc_span/src/span_encoding.rs

```diff
@@ -87,6 +87,106 @@ pub struct Span {
     ctxt_or_parent_or_marker: u16,
 }
 
+// Convenience structures for all span formats.
+#[derive(Clone, Copy)]
+struct InlineCtxt {
+    lo: u32,
+    len: u16,
+    ctxt: u16,
+}
+
+#[derive(Clone, Copy)]
+struct InlineParent {
+    lo: u32,
+    len_with_tag: u16,
+    parent: u16,
+}
+
+#[derive(Clone, Copy)]
+struct PartiallyInterned {
+    index: u32,
+    _marker1: u16,
+    ctxt: u16,
+}
+
+#[derive(Clone, Copy)]
+struct Interned {
+    index: u32,
+    _marker1: u16,
+    _marker2: u16,
+}
+
+enum Fmt<'a> {
+    InlineCtxt(&'a mut InlineCtxt),
+    InlineParent(&'a mut InlineParent),
+    PartiallyInterned(&'a mut PartiallyInterned),
+    Interned(&'a mut Interned),
+}
+
+impl InlineCtxt {
+    fn data(self) -> SpanData {
+        let len = self.len as u32;
+        debug_assert!(len <= MAX_LEN);
+        SpanData {
+            lo: BytePos(self.lo),
+            hi: BytePos(self.lo.debug_strict_add(len)),
+            ctxt: SyntaxContext::from_u32(self.ctxt as u32),
+            parent: None,
+        }
+    }
+    fn to_span(self) -> Span {
+        unsafe { transmute(self) }
+    }
+}
+
+impl InlineParent {
+    fn data(self) -> SpanData {
+        let len = (self.len_with_tag & !PARENT_TAG) as u32;
+        debug_assert!(len <= MAX_LEN);
+        SpanData {
+            lo: BytePos(self.lo),
+            hi: BytePos(self.lo.debug_strict_add(len)),
+            ctxt: SyntaxContext::root(),
+            parent: Some(LocalDefId { local_def_index: DefIndex::from_u32(self.parent as u32) }),
+        }
+    }
+    fn to_span(self) -> Span {
+        unsafe { transmute(self) }
+    }
+}
+
+impl PartiallyInterned {
+    fn data(self) -> SpanData {
+        SpanData {
+            ctxt: SyntaxContext::from_u32(self.ctxt as u32),
+            ..with_span_interner(|interner| interner.spans[self.index as usize])
+        }
+    }
+    fn to_span(self) -> Span {
+        unsafe { transmute(self) }
+    }
+}
+
+impl Interned {
+    fn data(self) -> SpanData {
+        with_span_interner(|interner| interner.spans[self.index as usize])
+    }
+    fn to_span(self) -> Span {
+        unsafe { transmute(self) }
+    }
+}
+
+impl Fmt<'_> {
```

Comment:
This name makes me think of `std::fmt`, but it's not related to that. Can we use a different name? `SpanKind`, maybe? Likewise for the `fmt` function.